topdown/http_test.go

```diff
@@ -340,3 +344,213 @@ func getTestServer() (baseURL string, teardownFn func()) {
 
 	return ts.URL, ts.Close
 }
+
+func getTLSTestServer() (ts *httptest.Server) {
+	mux := http.NewServeMux()
+	ts = httptest.NewUnstartedServer(mux)
+
+	mux.HandleFunc("/test", func(w http.ResponseWriter, req *http.Request) {
+		w.WriteHeader(http.StatusOK)
+	})
+
+	mux.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		// http.Redirect(w, req, "/test", http.StatusMovedPermanently)
+	})
+
+	return
+}
+
+func TestHTTPSClient(t *testing.T) {
+
+	const (
+		localClientCertFile  = "testdata/client-cert.pem"
+		localClientCert2File = "testdata/client-cert-2.pem"
+		localClientKeyFile   = "testdata/client-key.pem"
+		localCaFile          = "testdata/ca.pem"
+		localServerCertFile  = "testdata/server-cert.pem"
+		localServerKeyFile   = "testdata/server-key.pem"
+	)
+
+	caCertPEM, err := ioutil.ReadFile(localCaFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+	caPool := x509.NewCertPool()
+	if ok := caPool.AppendCertsFromPEM(caCertPEM); !ok {
+		t.Fatal("failed to parse CA cert")
+	}
+
+	cert, err := tls.LoadX509KeyPair(localServerCertFile, localServerKeyFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Set up Environment
+	clientCert, err := ReadCertFromFile(localClientCertFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.Setenv("CLIENT_CERT_ENV", string(clientCert))
+	if err != nil {
+		t.Fatal(err)
+	}
+	clientKey, err := ReadKeyFromFile(localClientKeyFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.Setenv("CLIENT_KEY_ENV", string(clientKey))
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.Setenv("CLIENT_CA_ENV", string(caCertPEM))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Replicating some of what happens in the server's HTTPS listener
+	s := getTLSTestServer()
+	s.TLS = &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		ClientCAs:    caPool,
+	}
+	s.StartTLS()
+	defer s.Close()
+
+	t.Run("Accessing Google with System Certs", func(t *testing.T) {
+
+		// Comment following line to run Internet Test using System Certs
```

Comment:
💭 Would we benefit from checking some ENV var that is set or not set here? Like, this could be something that is not set on travis so that we don't have flakey tests that depend on the state of google in our test runs...

Reply:
What's the goal of this test? I'd prefer to keep this off by default. I like to be able to work offline, however, that begs the question whether the test should exist in the first place.

Reply:
Removed the test

Reply:
And yes, using Env vars is something I did in previous open source projects to control testing. Travis has the necessary Env vars in place.

Real-world Internet tests (some call functional tests) are notorious difficult to reproduce and having them is good for the product in general.